File: app/src/main/java/wangdaye/com/geometricweather/weather/converters/OpenMeteoResultConverter.java
package wangdaye.com.geometricweather.weather.converters;

import android.content.Context;
import android.text.TextUtils;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.util.ArrayList;
import java.util.Date;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.TimeZone;

import wangdaye.com.geometricweather.R;
import wangdaye.com.geometricweather.common.basic.models.Location;
import wangdaye.com.geometricweather.common.basic.models.options.provider.WeatherSource;
import wangdaye.com.geometricweather.common.basic.models.weather.AirQuality;
import wangdaye.com.geometricweather.common.basic.models.weather.Astro;
import wangdaye.com.geometricweather.common.basic.models.weather.Base;
import wangdaye.com.geometricweather.common.basic.models.weather.Current;
import wangdaye.com.geometricweather.common.basic.models.weather.Daily;
import wangdaye.com.geometricweather.common.basic.models.weather.HalfDay;
import wangdaye.com.geometricweather.common.basic.models.weather.History;
import wangdaye.com.geometricweather.common.basic.models.weather.Hourly;
import wangdaye.com.geometricweather.common.basic.models.weather.Pollen;
import wangdaye.com.geometricweather.common.basic.models.weather.Precipitation;
import wangdaye.com.geometricweather.common.basic.models.weather.PrecipitationProbability;
import wangdaye.com.geometricweather.common.basic.models.weather.Temperature;
import wangdaye.com.geometricweather.common.basic.models.weather.UV;
import wangdaye.com.geometricweather.common.basic.models.weather.Weather;
import wangdaye.com.geometricweather.common.basic.models.weather.WeatherCode;
import wangdaye.com.geometricweather.common.basic.models.weather.Wind;
import wangdaye.com.geometricweather.common.basic.models.weather.WindDegree;
import wangdaye.com.geometricweather.common.utils.DisplayUtils;
import wangdaye.com.geometricweather.common.utils.helpers.LogHelper;
import wangdaye.com.geometricweather.weather.json.openmeteo.OpenMeteoAirQualityResult;
import wangdaye.com.geometricweather.weather.json.openmeteo.OpenMeteoLocationResults;
import wangdaye.com.geometricweather.weather.json.openmeteo.OpenMeteoWeatherResult;
import wangdaye.com.geometricweather.weather.services.WeatherService;

public class OpenMeteoResultConverter {

    @NonNull
    public static Location convert(@Nullable Location location, OpenMeteoLocationResults.Result result, WeatherSource weatherSource) {
        if (location != null
                && !TextUtils.isEmpty(location.getProvince())
                && !TextUtils.isEmpty(location.getCity())
                && !TextUtils.isEmpty(location.getDistrict())) {
            return new Location(
                    String.valueOf(result.id),
                    result.latitude,
                    result.longitude,
                    TimeZone.getTimeZone(result.timezone),
                    !TextUtils.isEmpty(result.country) ? result.country : result.countryCode,
                    location.getProvince(),
                    location.getCity(),
                    "",
                    null,
                    weatherSource,
                    false,
                    false,
                    !TextUtils.isEmpty(result.countryCode)
                            && (result.countryCode.equalsIgnoreCase("cn")
                            || result.countryCode.equalsIgnoreCase("hk")
                            || result.countryCode.equalsIgnoreCase("tw"))
            );
        } else {
            return new Location(
                    String.valueOf(result.id),
                    result.latitude,
                    result.longitude,
                    TimeZone.getTimeZone(result.timezone),
                    !TextUtils.isEmpty(result.country) ? result.country : result.countryCode,
                    TextUtils.isEmpty(result.admin2)
                            ? (TextUtils.isEmpty(result.admin1)
                                ? (TextUtils.isEmpty(result.admin3)
                                    ? (TextUtils.isEmpty(result.admin4)
                                        ? ""
                                        : result.admin4)
                                    : result.admin3)
                                : result.admin1)
                            : result.admin2,
                    result.name,
                    "",
                    null,
                    weatherSource,
                    false,
                    false,
                    !TextUtils.isEmpty(result.countryCode)
                            && (result.countryCode.equalsIgnoreCase("cn")
                            || result.countryCode.equalsIgnoreCase("hk")
                            || result.countryCode.equalsIgnoreCase("tw"))
            );
        }
    }

    @NonNull
    public static WeatherService.WeatherResultWrapper convert(Context context,
                                                              Location location,
                                                              OpenMeteoWeatherResult weatherResult/*,
                                                              OpenMeteoAirQualityResult airQualityResult*/) {
        try {
            List<Daily> initialDailyList = getInitialDailyList(context, weatherResult.daily);

            Map<String, Map<String, List<Hourly>>> hourlyByDate = new HashMap<>();
            List<Hourly> hourlyList = new ArrayList<>();
            for (int i = 0; i < weatherResult.hourly.time.length; ++i) {
                Hourly hourly = new Hourly(
                        new Date(weatherResult.hourly.time[i] * 1000),
                        weatherResult.hourly.time[i] * 1000,
                        weatherResult.hourly.isDay[i] > 0,
                        weatherResult.hourly.weatherCode[i] != null ? getWeatherText(context, weatherResult.hourly.weatherCode[i]) : null,
                        weatherResult.hourly.weatherCode[i] != null ? getWeatherCode(weatherResult.hourly.weatherCode[i]) : null,
                        new Temperature(
                                weatherResult.hourly.temperature[i] != null ? Math.round(weatherResult.hourly.temperature[i]) : null,
                                null,
                                null,
                                weatherResult.hourly.apparentTemperature[i] != null ? Math.round(weatherResult.hourly.apparentTemperature[i]) : null,
                                null,
                                null,
                                null
                        ),
                        new Precipitation(
                                weatherResult.hourly.precipitation[i], // TODO: It seems to not be the addition of the later
                                null,
                                weatherResult.hourly.rain[i], // Add showers[i]
                                weatherResult.hourly.snowfall[i],
                                null
                        ),
                        new PrecipitationProbability(
                                weatherResult.hourly.precipitationProbability[i] != null ? (float) weatherResult.hourly.precipitationProbability[i] : null,
                                null,
                                null,
                                null,
                                null
                        ),
                        new Wind(
                                weatherResult.hourly.windDirection[i] != null ? CommonConverter.getWindDirection(Float.valueOf(weatherResult.hourly.windDirection[i]), location.isChina()) : null,
                                weatherResult.hourly.windDirection[i] != null ? new WindDegree((float) weatherResult.hourly.windDirection[i], false) : null,
                                weatherResult.hourly.windSpeed[i],
                                weatherResult.hourly.windSpeed[i] != null ? CommonConverter.getWindLevel(
                                        context,
                                        weatherResult.hourly.windSpeed[i]
                                ) : null
                        ),
                        new AirQuality(null, null, null, null, null, null, null, null),
                        new Pollen(null, null, null, null, null, null, null, null, null, null, null, null),
                        new UV(weatherResult.hourly.uvIndex[i] != null ? Math.round(weatherResult.hourly.uvIndex[i]) : null, null, null)
                );

                // We shift by 6 hours the hourly date, otherwise night times (00:00 to 05:59) would be on the wrong day
                Date theDayAtMidnight = DisplayUtils.toTimezoneNoHour(new Date((weatherResult.hourly.time[i] - 6 * 3600) * 1000), location.getTimeZone());
                String theDayFormatted = DisplayUtils.getFormattedDate(theDayAtMidnight, location.getTimeZone(), "yyyyMMdd");
                if (!hourlyByDate.containsKey(theDayFormatted)) {
                    hourlyByDate.put(theDayFormatted, new HashMap<String, List<Hourly>>() {{
                        put("day", new ArrayList<>());
                        put("night", new ArrayList<>());
                    }});
                }
                if (weatherResult.hourly.time[i] < (theDayAtMidnight.getTime() / 1000) + 18 * 3600) {
                    // 06:00 to 17:59 is the day
                    hourlyByDate.get(theDayFormatted).get("day").add(hourly);
                } else {
                    // 18:00 to 05:59 is the night
                    hourlyByDate.get(theDayFormatted).get("night").add(hourly);
                }

                // Add to the app only if starts in the current hour
                if ((weatherResult.hourly.time[i]) >= (System.currentTimeMillis() / 1000) - 3600) {
                    hourlyList.add(hourly);
                }
            }
            List<Daily> dailyList = CommonConverter.completeDailyListWithHourlyList(initialDailyList, hourlyByDate, location.getTimeZone());

            Weather weather = new Weather(
                    new Base(
                            location.getCityId(),
                            System.currentTimeMillis(),
                            new Date(),
                            System.currentTimeMillis(),
                            new Date(),
                            System.currentTimeMillis()
                    ),
                    new Current(
                            getWeatherText(context, weatherResult.currentWeather.weatherCode),
                            getWeatherCode(weatherResult.currentWeather.weatherCode),
                            new Temperature(Math.round(weatherResult.currentWeather.temperature), null, null, null, null, null, null),
                            new Precipitation(null, null, null, null, null),
                            new Wind(
                                    weatherResult.currentWeather.windDirection != null ? CommonConverter.getWindDirection(Float.valueOf(weatherResult.currentWeather.windDirection), location.isChina()) : null,
                                    weatherResult.currentWeather.windDirection != null ? new WindDegree((float) weatherResult.currentWeather.windDirection, false) : null,
                                    weatherResult.currentWeather.windSpeed,
                                    weatherResult.currentWeather.windSpeed != null ? CommonConverter.getWindLevel(
                                            context,
                                            weatherResult.currentWeather.windSpeed
                                    ) : null
                            ),
                            null,
                            null,
                            null,
                            null,
                            null,
                            null,
                            null,
                            null,
                            null,
                            null
                    ),
                    new History(
                            new Date(weatherResult.daily.time[0] * 1000),
                            weatherResult.daily.time[0] * 1000,
                            weatherResult.daily.temperatureMax[0] != null ? Math.round(weatherResult.daily.temperatureMax[0]) : null,
                            weatherResult.daily.temperatureMin[0] != null ? Math.round(weatherResult.daily.temperatureMin[0]) : null
                    ),
                    dailyList,
                    hourlyList,
                    new ArrayList<>(),
                    new ArrayList<>()
            );
            return new WeatherService.WeatherResultWrapper(weather);
        } catch (Exception ignored) {
            return new WeatherService.WeatherResultWrapper(null);
        }
    }

    private static List<Daily> getInitialDailyList(Context context, OpenMeteoWeatherResult.Daily dailyResult) {
        List<Daily> initialDailyList = new ArrayList<>(dailyResult.time.length - 1);
        for (int i = 1; i < dailyResult.time.length; ++i) {
            Daily daily = new Daily(
                    new Date(dailyResult.time[i] * 1000),
                    dailyResult.time[i] * 1000,
                    new HalfDay(
                            null, // Completed later with hourly data
                            null, // Completed later with hourly data
                            null, // Completed later with hourly data
                            new Temperature(
                                    dailyResult.temperatureMax[i] != null ? Math.round(dailyResult.temperatureMax[i]) : null,
                                    null,
                                    null,
                                    dailyResult.apparentTemperatureMax[i] != null ? Math.round(dailyResult.apparentTemperatureMax[i]) : null,
                                    null,
                                    null,
                                    null
                            ),
                            null, // TODO: Complete with hourly data
                            null, // TODO: Complete with hourly data
                            null, // TODO: Complete with hourly data
                            null, // TODO: Complete with hourly data
                            null
                    ),
                    // For night temperature, we take the minTemperature from the following day
                    (i + 1) < dailyResult.time.length ? new HalfDay(
                            null, // Completed later with hourly data
                            null, // Completed later with hourly data
                            null, // Completed later with hourly data
                            new Temperature(
                                    dailyResult.temperatureMin[i + 1] != null ? Math.round(dailyResult.temperatureMin[i + 1]) : null,
                                    null,
                                    null,
                                    dailyResult.apparentTemperatureMin[i + 1] != null ? Math.round(dailyResult.apparentTemperatureMin[i + 1]) : null,
                                    null,
                                    null,
                                    null
                            ),
                            null, // TODO: Complete with hourly data
                            null, // TODO: Complete with hourly data
                            null, // TODO: Complete with hourly data
                            null, // TODO: Complete with hourly data
                            null
                    ) : null,
                    new Astro(
                            dailyResult.sunrise[i] != null ? new Date(dailyResult.sunrise[i] * 1000) : null,
                            dailyResult.sunset[i] != null ? new Date(dailyResult.sunset[i] * 1000) : null
                    ),
                    null,
                    null,
                    null, // TODO: Complete with hourly data
                    null, // TODO: Complete with hourly data
                    dailyResult.uvIndexMax[i] != null ? new UV(
                            Math.round(dailyResult.uvIndexMax[i]),
                            CommonConverter.getUVLevel(context, Math.round(dailyResult.uvIndexMax[i])),
                            null
                    ) : null,
                    dailyResult.sunrise[i] != null && dailyResult.sunset[i] != null ?
                            CommonConverter.getHoursOfDay(new Date(dailyResult.sunrise[i] * 1000), new Date(dailyResult.sunset[i] * 1000))
                            : null
            );
            initialDailyList.add(daily);
        }
        return initialDailyList;
    }

    private static String getWeatherText(Context context, Integer icon) {
        if (icon == null) {
            return null;
        }
        if (icon == 0) {
            return context.getString(R.string.weather_text_clear_sky);
        } else if (icon == 1) {
            return context.getString(R.string.weather_text_mainly_clear);
        } else if (icon == 2) {
            return context.getString(R.string.weather_text_partly_cloudy);
        } else if (icon == 3) {
            return context.getString(R.string.weather_text_overcast);
        } else if (icon == 45) {
            return context.getString(R.string.weather_text_fog);
        } else if (icon == 48) {
            return context.getString(R.string.weather_text_depositing_rime_fog);
        } else if (icon == 51) {
            return context.getString(R.string.weather_text_drizzle_light_intensity);
        } else if (icon == 53) {
            return context.getString(R.string.weather_text_drizzle_moderate_intensity);
        } else if (icon == 55) {
            return context.getString(R.string.weather_text_drizzle_dense_intensity);
        } else if (icon == 56) {
            return context.getString(R.string.weather_text_freezing_drizzle_light_intensity);
        } else if (icon == 57) {
            return context.getString(R.string.weather_text_freezing_drizzle_dense_intensity);
        } else if (icon == 61) {
            return context.getString(R.string.weather_text_rain_slight_intensity);
        } else if (icon == 63) {
            return context.getString(R.string.weather_text_rain_moderate_intensity);
        } else if (icon == 65) {
            return context.getString(R.string.weather_text_rain_heavy_intensity);
        } else if (icon == 66) {
            return context.getString(R.string.weather_text_freezing_rain_light_intensity);
        } else if (icon == 67) {
            return context.getString(R.string.weather_text_freezing_rain_heavy_intensity);
        } else if (icon == 71) {
            return context.getString(R.string.weather_text_snow_slight_intensity);
        } else if (icon == 73) {
            return context.getString(R.string.weather_text_snow_moderate_intensity);
        } else if (icon == 75) {
            return context.getString(R.string.weather_text_snow_heavy_intensity);
        } else if (icon == 77) {
            return context.getString(R.string.weather_text_snow_grains);
        } else if (icon == 80) {
            return context.getString(R.string.weather_text_rain_showers_slight);
        } else if (icon == 81) {
            return context.getString(R.string.weather_text_rain_showers_moderate);
        } else if (icon == 82) {
            return context.getString(R.string.weather_text_rain_showers_violent);
        } else if (icon == 85) {
            return context.getString(R.string.weather_text_snow_showers_slight);
        } else if (icon == 86) {
            return context.getString(R.string.weather_text_snow_showers_heavy);
        } else if (icon == 95) {
            return context.getString(R.string.weather_text_thunderstorm_slight_or_moderate);
        } else if (icon == 96) {
            return context.getString(R.string.weather_text_thunderstorm_with_slight_hail);
        } else if (icon == 99) {
            return context.getString(R.string.weather_text_thunderstorm_with_heavy_hail);
        } else {
            return null;
        }
    }

    private static WeatherCode getWeatherCode(Integer icon) {
        if (icon == null) {
            return null;
        }
        if (icon == 0 || icon == 1) { // Clear sky or Mainly clear
            return WeatherCode.CLEAR;
        } else if (icon == 2) { // Partly cloudy
            return WeatherCode.PARTLY_CLOUDY;
        } else if (icon == 3) { // Overcast
            return WeatherCode.CLOUDY;
        } else if (icon == 45 || icon == 48) { // Fog and depositing rime fog
            return WeatherCode.FOG;
        } else if (icon == 51 || icon == 53 || icon == 55 // Drizzle: Light, moderate, and dense intensity
                || icon == 56 || icon == 57 // Freezing Drizzle: Light and dense intensity
                || icon == 61 || icon == 63 || icon == 65 // Rain: Slight, moderate and heavy intensity
                || icon == 66 || icon == 67 // Freezing Rain: Light and heavy intensity
                || icon == 80 || icon == 81 || icon == 82) { // Rain showers: Slight, moderate, and violent
            return WeatherCode.RAIN;
        } else if (icon == 71 || icon == 73 || icon == 75 // Snow fall: Slight, moderate, and heavy intensity
                || icon == 85 || icon == 86) { // Snow showers slight and heavy
            return WeatherCode.SNOW;
        } else if (icon == 77) { // Snow grains
            return WeatherCode.SLEET;
        } else if (icon == 95 // Thunderstorm: Slight or moderate
                || icon == 96 || icon == 99) { // Thunderstorm with slight and heavy hail
            return WeatherCode.THUNDERSTORM;
        } else {
            return WeatherCode.CLOUDY;
        }
    }
}

File: app/src/main/java/org/breezyweather/sources/openmeteo/OpenMeteoResultConverter.kt
/**
 * This file is part of Breezy Weather.
 *
 * Breezy Weather is free software: you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as published by the
 * Free Software Foundation, version 3 of the License.
 *
 * Breezy Weather is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public
 * License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with Breezy Weather. If not, see <https://www.gnu.org/licenses/>.
 */

package org.breezyweather.sources.openmeteo

import android.content.Context
import breezyweather.domain.location.model.Location
import breezyweather.domain.location.model.LocationAddressInfo
import breezyweather.domain.weather.model.AirQuality
import breezyweather.domain.weather.model.DailyCloudCover
import breezyweather.domain.weather.model.DailyDewPoint
import breezyweather.domain.weather.model.DailyPressure
import breezyweather.domain.weather.model.DailyRelativeHumidity
import breezyweather.domain.weather.model.DailyVisibility
import breezyweather.domain.weather.model.Minutely
import breezyweather.domain.weather.model.Pollen
import breezyweather.domain.weather.model.Precipitation
import breezyweather.domain.weather.model.PrecipitationProbability
import breezyweather.domain.weather.model.UV
import breezyweather.domain.weather.model.Wind
import breezyweather.domain.weather.reference.WeatherCode
import breezyweather.domain.weather.wrappers.AirQualityWrapper
import breezyweather.domain.weather.wrappers.CurrentWrapper
import breezyweather.domain.weather.wrappers.DailyWrapper
import breezyweather.domain.weather.wrappers.HalfDayWrapper
import breezyweather.domain.weather.wrappers.HourlyWrapper
import breezyweather.domain.weather.wrappers.PollenWrapper
import breezyweather.domain.weather.wrappers.TemperatureWrapper
import breezyweather.domain.weather.wrappers.WeatherWrapper
import org.breezyweather.R
import org.breezyweather.common.extensions.plus
import org.breezyweather.common.extensions.toCalendarWithTimeZone
import org.breezyweather.common.extensions.toDate
import org.breezyweather.sources.openmeteo.json.OpenMeteoAirQualityHourly
import org.breezyweather.sources.openmeteo.json.OpenMeteoAirQualityResult
import org.breezyweather.sources.openmeteo.json.OpenMeteoLocationResult
import org.breezyweather.sources.openmeteo.json.OpenMeteoWeatherCurrent
import org.breezyweather.sources.openmeteo.json.OpenMeteoWeatherDaily
import org.breezyweather.sources.openmeteo.json.OpenMeteoWeatherHourly
import org.breezyweather.sources.openmeteo.json.OpenMeteoWeatherMinutely
import org.breezyweather.sources.openmeteo.json.OpenMeteoWeatherResult
import java.util.Calendar
import java.util.Date
import java.util.TimeZone
import kotlin.math.roundToInt
import kotlin.time.Duration.Companion.minutes
import kotlin.time.Duration.Companion.seconds

internal fun getCurrent(
    current: OpenMeteoWeatherCurrent?,
    context: Context,
): CurrentWrapper? {
    if (current == null) return null

    return CurrentWrapper(
        weatherText = getWeatherText(context, current.weatherCode),
        weatherCode = getWeatherCode(current.weatherCode),
        temperature = TemperatureWrapper(
            temperature = current.temperature,
            feelsLike = current.apparentTemperature
        ),
        wind = Wind(
            degree = current.windDirection,
            speed = current.windSpeed,
            gusts = current.windGusts
        ),
        uV = UV(index = current.uvIndex),
        relativeHumidity = current.relativeHumidity?.toDouble(),
        dewPoint = current.dewPoint,
        pressure = current.pressureMsl,
        cloudCover = current.cloudCover,
        visibility = current.visibility
    )
}

internal fun getDailyList(
    dailyResult: OpenMeteoWeatherDaily?,
    location: Location,
): List<DailyWrapper>? {
    if (dailyResult == null) return null

    val dailyList: MutableList<DailyWrapper> = ArrayList(dailyResult.time.size - 1)
    for (i in 0 until dailyResult.time.size - 1) {
        val theDayWithDstFixed = dailyResult.time[i].seconds.inWholeMilliseconds.toDate()
            .toCalendarWithTimeZone(location.timeZone)
            .apply {
                add(Calendar.HOUR_OF_DAY, 1)
                set(Calendar.HOUR_OF_DAY, 0)
            }.time
        val daily = DailyWrapper(
            date = theDayWithDstFixed,
            day = HalfDayWrapper(
                temperature = TemperatureWrapper(
                    temperature = dailyResult.temperatureMax?.getOrNull(i),
                    feelsLike = dailyResult.apparentTemperatureMax?.getOrNull(i)
                )
            ),
            night = HalfDayWrapper(
                temperature = TemperatureWrapper(
                    // For night temperature, we take the minTemperature from the following day
                    temperature = dailyResult.temperatureMin?.getOrNull(i + 1),
                    feelsLike = dailyResult.apparentTemperatureMin?.getOrNull(i + 1)
                )
            ),
            uV = UV(index = dailyResult.uvIndexMax?.getOrNull(i)),
            sunshineDuration = dailyResult.sunshineDuration?.getOrNull(i)?.div(3600),
            relativeHumidity = DailyRelativeHumidity(
                average = dailyResult.relativeHumidityMean?.getOrNull(i)?.toDouble(),
                max = dailyResult.relativeHumidityMax?.getOrNull(i)?.toDouble(),
                min = dailyResult.relativeHumidityMin?.getOrNull(i)?.toDouble()
            ),
            dewPoint = DailyDewPoint(
                average = dailyResult.dewPointMean?.getOrNull(i),
                max = dailyResult.dewPointMax?.getOrNull(i),
                min = dailyResult.dewPointMin?.getOrNull(i)
            ),
            pressure = DailyPressure(
                average = dailyResult.pressureMslMean?.getOrNull(i),
                max = dailyResult.pressureMslMax?.getOrNull(i),
                min = dailyResult.pressureMslMin?.getOrNull(i)
            ),
            cloudCover = DailyCloudCover(
                average = dailyResult.cloudCoverMean?.getOrNull(i),
                max = dailyResult.cloudCoverMax?.getOrNull(i),
                min = dailyResult.cloudCoverMin?.getOrNull(i)
            ),
            visibility = DailyVisibility(
                average = dailyResult.visibilityMean?.getOrNull(i),
                max = dailyResult.visibilityMax?.getOrNull(i),
                min = dailyResult.visibilityMin?.getOrNull(i)
            )
        )
        dailyList.add(daily)
    }
    return dailyList
}

internal fun getHourlyList(
    context: Context,
    hourlyResult: OpenMeteoWeatherHourly?,
): List<HourlyWrapper>? {
    if (hourlyResult == null) return null

    val hourlyList = mutableListOf<HourlyWrapper>()
    for (i in hourlyResult.time.indices) {
        hourlyList.add(
            HourlyWrapper(
                date = hourlyResult.time[i].seconds.inWholeMilliseconds.toDate(),
                isDaylight = if (hourlyResult.isDay?.getOrNull(i) != null) hourlyResult.isDay[i] > 0 else null,
                weatherText = getWeatherText(context, hourlyResult.weatherCode?.getOrNull(i)),
                weatherCode = getWeatherCode(hourlyResult.weatherCode?.getOrNull(i)),
                temperature = TemperatureWrapper(
                    temperature = hourlyResult.temperature?.getOrNull(i),
                    feelsLike = hourlyResult.apparentTemperature?.getOrNull(i)
                ),
                precipitation = Precipitation(
                    total = hourlyResult.precipitation?.getOrNull(i),
                    rain = hourlyResult.rain?.getOrNull(i) + hourlyResult.showers?.getOrNull(i),
                    snow = hourlyResult.snowfall?.getOrNull(i)?.times(10) // convert cm -> mm
                ),
                precipitationProbability = PrecipitationProbability(
                    total = hourlyResult.precipitationProbability?.getOrNull(i)?.toDouble()
                ),
                wind = Wind(
                    degree = hourlyResult.windDirection?.getOrNull(i)?.toDouble(),
                    speed = hourlyResult.windSpeed?.getOrNull(i),
                    gusts = hourlyResult.windGusts?.getOrNull(i)
                ),
                uV = UV(index = hourlyResult.uvIndex?.getOrNull(i)),
                relativeHumidity = hourlyResult.relativeHumidity?.getOrNull(i)?.toDouble(),
                dewPoint = hourlyResult.dewPoint?.getOrNull(i),
                pressure = hourlyResult.pressureMsl?.getOrNull(i),
                cloudCover = hourlyResult.cloudCover?.getOrNull(i),
                visibility = hourlyResult.visibility?.getOrNull(i)?.toDouble()
            )
        )
    }
    return hourlyList
}

internal fun getAirQuality(
    hourlyAirQualityResult: OpenMeteoAirQualityHourly?,
): AirQualityWrapper? {
    if (hourlyAirQualityResult == null) return null

    val airQualityHourly = mutableMapOf<Date, AirQuality>()
    for (i in hourlyAirQualityResult.time.indices) {
        airQualityHourly[hourlyAirQualityResult.time[i].seconds.inWholeMilliseconds.toDate()] = AirQuality(
            pM25 = hourlyAirQualityResult.pm25?.getOrNull(i),
            pM10 = hourlyAirQualityResult.pm10?.getOrNull(i),
            sO2 = hourlyAirQualityResult.sulphurDioxide?.getOrNull(i),
            nO2 = hourlyAirQualityResult.nitrogenDioxide?.getOrNull(i),
            o3 = hourlyAirQualityResult.ozone?.getOrNull(i),
            cO = hourlyAirQualityResult.carbonMonoxide?.getOrNull(i)?.div(1000.0)
        )
    }
    return AirQualityWrapper(
        hourlyForecast = airQualityHourly
    )
}

internal fun getPollen(
    hourlyAirQualityResult: OpenMeteoAirQualityHourly?,
): PollenWrapper? {
    if (hourlyAirQualityResult == null) return null

    val pollenHourly = mutableMapOf<Date, Pollen>()
    for (i in hourlyAirQualityResult.time.indices) {
        pollenHourly[hourlyAirQualityResult.time[i].seconds.inWholeMilliseconds.toDate()] = Pollen(
            alder = hourlyAirQualityResult.alderPollen?.getOrNull(i)?.roundToInt(),
            birch = hourlyAirQualityResult.birchPollen?.getOrNull(i)?.roundToInt(),
            grass = hourlyAirQualityResult.grassPollen?.getOrNull(i)?.roundToInt(),
            mugwort = hourlyAirQualityResult.mugwortPollen?.getOrNull(i)?.roundToInt(),
            olive = hourlyAirQualityResult.olivePollen?.getOrNull(i)?.roundToInt(),
            ragweed = hourlyAirQualityResult.ragweedPollen?.getOrNull(i)?.roundToInt()
        )
    }
    return PollenWrapper(
        hourlyForecast = pollenHourly
    )
}

internal fun getMinutelyList(
    minutelyFifteen: OpenMeteoWeatherMinutely?,
): List<Minutely>? {
    if (minutelyFifteen?.time == null || minutelyFifteen.time.isEmpty()) return null

    val currentMinutelyIndex = minutelyFifteen.time.indexOfFirst {
        it.seconds.inWholeMilliseconds >= (Date().time - 15.minutes.inWholeMilliseconds)
    }
    val maxMinutelyIndex = minOf(currentMinutelyIndex + 8, minutelyFifteen.time.size - 1)
    val precipitationMinutely = minutelyFifteen.precipitation?.slice(currentMinutelyIndex until maxMinutelyIndex)
    // val precipitationProbabilityMinutely =
    //     minutelyFifteen.precipitationProbability?.slice(currentMinutelyIndex until maxMinutelyIndex)
    return minutelyFifteen.time.slice(currentMinutelyIndex until maxMinutelyIndex)
        // 2 hours
        .mapIndexed { i, time ->
            Minutely(
                date = time.seconds.inWholeMilliseconds.toDate(),
                minuteInterval = 15,
                // mm/15 min -> mm/h
                precipitationIntensity = precipitationMinutely?.getOrNull(i)?.times(4)
                /*if (precipitationProbabilityMinutely?.getOrNull(i) != null &&
                    precipitationProbabilityMinutely[i]!! > 30
                ) {
                    precipitationMinutely?.getOrNull(i)?.times(4) // mm/15 min -> mm/h
                } else {
                    null
                }*/
            )
        }
}

private fun getWeatherText(
    context: Context,
    icon: Int?,
): String? {
    return when (icon) {
        null -> null
        0 -> context.getString(R.string.openmeteo_weather_text_clear_sky)
        1 -> context.getString(R.string.openmeteo_weather_text_mainly_clear)
        2 -> context.getString(R.string.openmeteo_weather_text_partly_cloudy)
        3 -> context.getString(R.string.openmeteo_weather_text_overcast)
        45 -> context.getString(R.string.openmeteo_weather_text_fog)
        48 -> context.getString(R.string.openmeteo_weather_text_depositing_rime_fog)
        51 -> context.getString(R.string.openmeteo_weather_text_drizzle_light_intensity)
        53 -> context.getString(R.string.openmeteo_weather_text_drizzle_moderate_intensity)
        55 -> context.getString(R.string.openmeteo_weather_text_drizzle_dense_intensity)
        56 -> context.getString(R.string.openmeteo_weather_text_freezing_drizzle_light_intensity)
        57 -> context.getString(R.string.openmeteo_weather_text_freezing_drizzle_dense_intensity)
        61 -> context.getString(R.string.openmeteo_weather_text_rain_slight_intensity)
        63 -> context.getString(R.string.openmeteo_weather_text_rain_moderate_intensity)
        65 -> context.getString(R.string.openmeteo_weather_text_rain_heavy_intensity)
        66 -> context.getString(R.string.openmeteo_weather_text_freezing_rain_light_intensity)
        67 -> context.getString(R.string.openmeteo_weather_text_freezing_rain_heavy_intensity)
        71 -> context.getString(R.string.openmeteo_weather_text_snow_slight_intensity)
        73 -> context.getString(R.string.openmeteo_weather_text_snow_moderate_intensity)
        75 -> context.getString(R.string.openmeteo_weather_text_snow_heavy_intensity)
        77 -> context.getString(R.string.openmeteo_weather_text_snow_grains)
        80 -> context.getString(R.string.openmeteo_weather_text_rain_showers_slight)
        81 -> context.getString(R.string.openmeteo_weather_text_rain_showers_moderate)
        82 -> context.getString(R.string.openmeteo_weather_text_rain_showers_violent)
        85 -> context.getString(R.string.openmeteo_weather_text_snow_showers_slight)
        86 -> context.getString(R.string.openmeteo_weather_text_snow_showers_heavy)
        95 -> context.getString(R.string.openmeteo_weather_text_thunderstorm_slight_or_moderate)
        96 -> context.getString(R.string.openmeteo_weather_text_thunderstorm_with_slight_hail)
        99 -> context.getString(R.string.openmeteo_weather_text_thunderstorm_with_heavy_hail)
        else -> null
    }
}

private fun getWeatherCode(
    icon: Int?,
): WeatherCode? {
    return when (icon) {
        null -> null
        0, 1 -> WeatherCode.CLEAR // Clear sky or Mainly clear
        2 -> WeatherCode.PARTLY_CLOUDY // Partly cloudy
        3 -> WeatherCode.CLOUDY // Overcast
        45, 48 -> WeatherCode.FOG // Fog and depositing rime fog
        51, 53, 55, // Drizzle: Light, moderate, and dense intensity
        56, 57, // Freezing Drizzle: Light and dense intensity
        61, 63, 65, // Rain: Slight, moderate and heavy intensity
        66, 67, // Freezing Rain: Light and heavy intensity
        80, 81, 82, // Rain showers: Slight, moderate, and violent
        -> WeatherCode.RAIN
        71, 73, 75, // Snow fall: Slight, moderate, and heavy intensity
        85, 86, // Snow showers slight and heavy
        -> WeatherCode.SNOW
        77 -> WeatherCode.SLEET // Snow grains
        95, 96, 99 -> WeatherCode.THUNDERSTORM // Thunderstorm with slight and heavy hail
        else -> null
    }
}

/**
 * Functions for debugging purposes (tracking NPE)
 */
// Sometimes used in dev to make some null-safety checks
// TODO: Should be moved to its own DebugWeatherService
internal fun debugConvert(
    context: Context,
    weatherResult: OpenMeteoWeatherResult,
    airQualityResult: OpenMeteoAirQualityResult,
): WeatherWrapper {
    val dailyList = mutableListOf<DailyWrapper>()
    if (weatherResult.daily != null) {
        for (i in 1 until weatherResult.daily.time.size) {
            val daily = DailyWrapper(date = weatherResult.daily.time[i].seconds.inWholeMilliseconds.toDate())
            dailyList.add(daily)
        }
    }

    val hourlyList = mutableListOf<HourlyWrapper>()
    if (weatherResult.hourly != null) {
        for (i in weatherResult.hourly.time.indices) {
            // Add to the app only if starts in the current hour
            if (weatherResult.hourly.time[i] >= System.currentTimeMillis() / 1000 - 3600) {
                hourlyList.add(HourlyWrapper(date = weatherResult.hourly.time[i].seconds.inWholeMilliseconds.toDate()))
            }
        }
    }

    return WeatherWrapper(
        dailyForecast = dailyList,
        hourlyForecast = hourlyList
    )
}
